// app-controller/index.js
var _ = require('lodash');
var log = require('app-log');

var responders = {
	'POST': responsePOST,
	'GET': responseGET,
	'PUT': responsePUT,
	'DELETE': responsePUT,
	'PATCH': responsePUT
};

function responsePOST (data, res) {
	return res.status(201).json(data);
}

function responseGET (data, res) {
	if (!data) {
		return res.status(404).end();
	}

	return res.status(200).json(data);
}

function responsePUT (data, res) {
	if (!data) {
		return res.status(204).end();
	}

	return res.status(200).json(data);
}

// Wrap function with autoresponse with promises and error handling
function wrap (action) {
	if (!_.isFunction(action)) {
		throw new Error('Controller action must be a function but ' + typeof action + ' is given');
	}

	return function wrapAction (req, res) {

		// Gather all possible params to a single object
		var params = _.extend({}, req.params, req.query, req.body);

		try {
			// Execute handler
			var result = action(params, req, res);

			// Check if result is promise then do chaining
			if (!_.isUndefined(result) && _.isFunction(result.done)) {
				result.done(onFulfilled, onRejected);
			} else {
				onFulfilled(result);
			}

		} catch (err) {
			onRejected(err);
		}

		// Success handler
		function onFulfilled (data) {
			// If headers are already sent, just noop
			if (res.headersSent) {
				return;
			}

			// Response with json
			responders[req.method](data, res);
		}

		// Error handler
		function onRejected (err) {
			errorHandler(req, res, err);
		}
	}
}

// Error handler
function errorHandler (req, res, err) {
	// In case of failed app-validation
	if (err && err.errors) {
		res.status(400).json(err);
	} else {
		log.error(err);
		res.status(500).end();
	}
}

// Set global error handler (for all actions)
function setErrorHandler (handler) {
	if (handler && !_.isFunction(handler)) {
		throw new Error('Error handler should be a function');
	}

	errorHandler = handler;
}

// Set controller logger (uses app-log interface)
function setLogger (logger) {
	log = logger;
}

// Wrap function or functions object
function wrapAll (controller) {
	if (_.isFunction(controller)) {
		return wrap(controller);
	}

	if (!_.isObject(controller)) {
		throw new Error('Controller must be function or object of functions');
	}

	var result = {};
	for (var name in controller) {
		result[name] = wrap(controller[name]);
	}

	return result;
}

module.exports = wrapAll;
module.exports.setErrorHandler = setErrorHandler;
module.exports.setLogger = setLogger;
